### PR TITLE
Moved Semaphore.cs and Semaphore.Windows.cs to shared

### PR DESCRIPTION
**Pull request by Anipik**
Related CorertPR :- https://github.com/dotnet/corert/pull/5986

### CODE REVIEW

File: src/System.Private.CoreLib/src/Microsoft/Win32/Win32Native.cs [hunk 1]
```diff
@@ -165,7 +165,6 @@ internal static class Win32Native
 
         // Note that you may need to specify the SYNCHRONIZE bit as well
         // to be able to open a synchronization primitive.
-        internal const int SEMAPHORE_MODIFY_STATE = 0x00000002;
         internal const int EVENT_MODIFY_STATE = 0x00000002;
         internal const int MUTEX_MODIFY_STATE = 0x00000001;
```
Comment (jkotas) [inline src/System.Private.CoreLib/src/Microsoft/Win32/Win32Native.cs]: `MUTEX_MODIFY_STATE` looks unused too

**Comment by Anipik** [comment]
@dotnet-bot test CentOS7.1 x64 Checked Innerloop Build and Test

**Comment by Anipik** [comment]
@dotnet-bot test CentOS7.1 x64 Checked Innerloop Build and Test

File: src/System.Private.CoreLib/src/Microsoft/Win32/Win32Native.cs (hunk 1 above)
Comment (Anipik) [reply src/System.Private.CoreLib/src/Microsoft/Win32/Win32Native.cs]: I will have another PR to move waithandle.cs to shared. I will remove this in that

File: src/System.Private.CoreLib/shared/System/Threading/Semaphore.Windows.cs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+            SafeWaitHandle = handle;
+        }
+
+        private void CreateSemaphoreCore(int initialCount, int maximumCount, string name, out bool createdNew)
+        {
+            Debug.Assert(initialCount >= 0);
+            Debug.Assert(maximumCount >= 1);
+            Debug.Assert(initialCount <= maximumCount);
+
+#if PLATFORM_UNIX
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Threading/Semaphore.Windows.cs]: Can we make this instead `#if !PLATFORM_WINDOWS`?  I just got rid of the last uses of PLATFORM_UNIX in corelib:
https://github.com/dotnet/coreclr/pull/18411

File: src/System.Private.CoreLib/shared/System/Threading/Semaphore.Windows.cs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            if (name != null)
+                throw new PlatformNotSupportedException(SR.PlatformNotSupported_NamedSynchronizationPrimitives);
+#endif
+            SafeWaitHandle myHandle = Interop.Kernel32.CreateSemaphoreEx(IntPtr.Zero, initialCount, maximumCount, name, 0, AccessRights);
+
+            int errorCode = Marshal.GetLastWin32Error();
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Threading/Semaphore.Windows.cs]: Why was this moved to not be guarded by the IsInvalid check?

File: src/System.Private.CoreLib/shared/System/Threading/Semaphore.Windows.cs [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+            //Pass false to OpenSemaphore to prevent inheritedHandles
+            SafeWaitHandle myHandle = Interop.Kernel32.OpenSemaphore(AccessRights, false, name);
+
+            if (myHandle.IsInvalid)
+            {
+                result = null;
+                int errorCode = Marshal.GetLastWin32Error();
+
+                if (errorCode == Interop.Errors.ERROR_FILE_NOT_FOUND || errorCode ==  Interop.Errors.ERROR_INVALID_NAME)
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Threading/Semaphore.Windows.cs]: Nit: there's an extra space after the second ==

File: src/System.Private.CoreLib/shared/System/Threading/Semaphore.cs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        {
+            VerifyCounts(initialCount, maximumCount);
+
+            bool createdNew;
+            CreateSemaphoreCore(initialCount, maximumCount, name, out createdNew);
+        }
+
+        public Semaphore(int initialCount, int maximumCount, string name, out bool createdNew)
+        {
+            VerifyCounts(initialCount, maximumCount);
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Threading/Semaphore.cs]: Nit: I'd suggest changing VerifyCounts to be named something more like ValidateArguments.  It might just be me, but when I see "VerifyCounts", I think it's doing something more with the internal state of the class.

File: src/System.Private.CoreLib/shared/System/Threading/Semaphore.cs [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        // Win32 only takes maximum count of Int32.MaxValue
+        public Semaphore(int initialCount, int maximumCount) : this(initialCount, maximumCount, null) { }
+
+        public Semaphore(int initialCount, int maximumCount, string name)
+        {
+            VerifyCounts(initialCount, maximumCount);
+
+            bool createdNew;
+            CreateSemaphoreCore(initialCount, maximumCount, name, out createdNew);
+        }
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Threading/Semaphore.cs]: Rather than refactoring the argument validation into its own method, you could instead just have it in the other ctor and make this constructor be:
```C#
public Semaphore(int initialCount, int maximumCount, string name) : this(initialCount, maximumCount, name, out bool createdNew)
{
}
```
That will require using C# 7.3, though. @jkotas, are we using 7.3 in coreclr yet?

Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/Threading/Semaphore.cs]: Actually, it can be done with 7.2 (which we definitely allow) with just:
```C#
public Semaphore(int initialCount, int maximumCount, string name) :
    this(initialCount, maximumCount, name, out _)
{
}
```

File: src/System.Private.CoreLib/shared/System/Threading/Semaphore.Windows.cs (hunk 3 above)
Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/Threading/Semaphore.Windows.cs]: I see, it occurred in two different methods, in one before the if and in the other guarded by the if.  I suggest we move this to be inside the if block so that we don't call GetLastWin32Error unless we need to.

**Comment by Anipik** [comment]
@stephentoub I will make all these changes in this https://github.com/dotnet/corert/pull/5990

File: src/System.Private.CoreLib/shared/System/Threading/Semaphore.Windows.cs (hunk 3 above)
Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/Threading/Semaphore.Windows.cs]: CreateSemaphoreEx sets last error even when it returns valid handle. I think that it fine to always call GetLastWin32Error before the if block.

Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/Threading/Semaphore.Windows.cs]: But we don't do anything with it. Does GetLastWin32Error clear the error and that's why it's good to call regardless of whether the handle is valid?

Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/Threading/Semaphore.Windows.cs]: Oh, nevermind, we do do something with it.  Ignore me.
